### PR TITLE
[MLX] Qwen3.5 MoE ondevice sampling

### PR DESCRIPTION
**Pull request by kiymetakdemir**
Summary
Lets the MLX-exported Qwen3.5 MoE model sample the next token on-device instead of returning logits for host-side sampling. Sampling is opt-in at export (--sample); temperature, top_p, and seed are runtime inputs, and the runner increments the seed per token. Measured ~17% higher decode throughput.

  Changes
  - export.py --sample flag wraps the model in SamplingHead so forward(tokens, input_pos, temperature, top_p, seed) → int64 token; records a use_sampling constant-method flag. Non-sample export unchanged.
  - qwen35_moe_engine.cpp reads use_sampling from metadata; when set, consumes the token id directly instead of logits_to_token, feeds the scalar inputs in prefill/decode, and manages the per-token seed schedule. top_k is still rejected; top_p/seed are rejected on non-sample models. 
  - main.cpp --top_p / --seed flags wired into SamplingConfig.
  - run.py detects use_sampling, mirrors the same seed schedule for parity, and rejects top_p/seed on non-sample models (matching the C++ runner). --top-p / --seed flags added.

  Verification
  - Real model (C++ runner): coherent, constraint-following generation; same --seed reproducible; --temperature 0 deterministic.
  - +17% decode throughput vs host sampling; prefill unchanged.

### CODE REVIEW

File: examples/models/qwen3_5_moe/export.py [hunk 1]
```diff
@@ -733,6 +733,25 @@ def _clean_forward(self, tokens, input_pos):
     model.forward = types.MethodType(_clean_forward, model)
 
 
+class _MLXSampleWrapper(nn.Module):
```
Comment (metascroy) [inline examples/models/qwen3_5_moe/export.py:736]: Can we just use SamplingHead directly? What does the wrapper do extra?

File: examples/models/qwen3_5_moe/export.py [hunk 2]
```diff
@@ -749,16 +768,33 @@ def _export_mlx(model, config, args):
...
+
     example_tokens = torch.tensor([[0, 1]], dtype=torch.long)
     example_input_pos = torch.tensor([0, 1], dtype=torch.long)
     seq_dim = Dim("seq_len", min=1, max=config.max_seq_len - 1)
-    dynamic_shapes = ({1: seq_dim}, {0: seq_dim})
+
+    if sample:
+        # forward(tokens, input_pos, temperature, top_p, seed) -> token id.
+        # Scalars are static (None in dynamic_shapes); only the seq dim is dynamic.
+        model = _MLXSampleWrapper(model)
```
Comment (metascroy) [inline examples/models/qwen3_5_moe/export.py:780]: Use SamplingHead directly?

File: examples/models/qwen3_5_moe/main.cpp [hunk 3]
```diff
@@ -37,6 +37,16 @@ DEFINE_string(
...
     "Path to file containing prompt text (overrides --prompt).");
 DEFINE_double(temperature, 0.8, "Sampling temperature (0 = greedy).");
+DEFINE_double(
+    top_p,
+    1.0,
+    "Nucleus sampling top_p in (0, 1]; 1.0 = off. Requires a model exported "
+    "with --sample (MLX on-device sampling).");
+DEFINE_int64(
+    seed,
+    0,
```
Comment (metascroy) [inline examples/models/qwen3_5_moe/main.cpp]: Can default mean we generate base seed by calling random int from c++ library?

**Comment by metascroy** [comment]
@claude review this code

File: examples/models/qwen3_5_moe/export.py (hunk 1 above)
Comment (kiymetakdemir) [reply examples/models/qwen3_5_moe/export.py:736]: It is because temperature, top_p, seed are keyword only arguments, the wrapper maps them into positional args. I can pass them to export's kwargs instead but it also makes dynamic_shapes a structured args/kwargs dict rather than a tuple.

File: examples/models/qwen3_5_moe/main.cpp (hunk 3 above)
Comment (kiymetakdemir) [reply examples/models/qwen3_5_moe/main.cpp]: Added this.

**Review by metascroy — APPROVED** [review]

File: examples/models/qwen3_5_moe/export.py [hunk 4]
```diff
@@ -1182,6 +1219,13 @@ def main():  # noqa: C901
         action="store_true",
         help="Disable split-K (flash-decoding) SDPA for decode; use tiled SDPA instead.",
     )
+    parser.add_argument(
+        "--sample",
+        action="store_true",
+        help="MLX only: sample the next token on-device (Gumbel-max with "
```
Comment (digantdesai) [inline examples/models/qwen3_5_moe/export.py:1225]: Can we make both cuda/mlx UX similar? https://github.com/pytorch/executorch/pull/18844
